File: src/test/java/com/conference/website/service/E10_TagServiceTest.java
package com.conference.website.service;

import com.conference.website.dto.CreateTagsRequest;
import com.conference.website.dto.TagDto;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

@SpringBootTest
@Transactional
class E10_TagServiceTest {

    @Autowired
    private TagService tagService;

    @Test
    void shouldCreateTag() {
        //Arrange, Act
        List<TagDto> createdTags = tagService.createTags(
           new CreateTagsRequest(List.of("Java")));

        //Assert
        assertEquals(2, createdTags.size());
        var firstTag = createdTags.getFirst();
        assertEquals("java", firstTag.name());
        assertNotNull(firstTag.id());
    }



















    @Test
    void shouldRejectDuplicateTagNames() {
        List<TagDto> createdTags = tagService
           .createTags(new CreateTagsRequest(
              List.of("java", "kotlin")));

        assertThat(createdTags)
           .hasSize(3)
                .allSatisfy(tag ->
                   assertThat(tag.id()).isNotNull())
                .extracting(TagDto::name)
           .containsExactlyInAnyOrder("java", "kotlin") ;

        assertThatThrownBy(() -> tagService
           .createTags(new CreateTagsRequest(List.of("Java"))))
                .isInstanceOf(BadRequestException.class)
                .hasMessageContaining("Tag already exists")
                .hasMessageContaining("java");
    }

}

File: src/test/kotlin/com/conference/website/service/E10_TagServiceTest.kt
package com.conference.website.service

import com.conference.website.dto.CreateTagsRequest
import com.conference.website.dto.TagDto
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@Transactional
class E10_TagServiceTest {

    @Autowired
    private lateinit var tagService: TagService

    @Test
    fun shouldCreateTag() {
        //Arrange, Act
        val createdTags: List<TagDto> = tagService.createTags(
            CreateTagsRequest(listOf("Java"))
        )

        //Assert
        assertEquals(2, createdTags.size)
        val firstTag = createdTags.first()
        assertEquals("java", firstTag.name)
        assertNotNull(firstTag.id)
    }



















    @Test
    fun shouldRejectDuplicateTagNames() {
        val createdTags: List<TagDto> = tagService
            .createTags(
                CreateTagsRequest(
                    listOf("java", "kotlin")
                )
            )

        assertThat(createdTags)
            .hasSize(3)
            .allSatisfy { tag -> assertThat(tag.id).isNotNull() }
            .extracting<String>(TagDto::name)
            .containsExactlyInAnyOrder("java", "kotlin")

        assertThatThrownBy {
            tagService.createTags(CreateTagsRequest(listOf("Java")))
        }
            .isInstanceOf(BadRequestException::class.java)
            .hasMessageContaining("Tag already exists")
            .hasMessageContaining("java")
    }
}
